source/core_mqtt.c

```diff
@@ -2712,14 +2797,24 @@ MQTTStatus_t MQTT_Connect( MQTTContext_t * pContext,
 
     if( status == MQTTSuccess )
     {
-        MQTT_PRE_SEND_HOOK( pContext );
+        MQTT_PRE_STATE_UPDATE_HOOK( pContext );
+        
+        stateUpdateHookExecuted = true;
 
+        connectStatus=pContext->connectStatus;
+
+        if( connectStatus != MQTTNotConnected)
+        {
+            status = (connectStatus==MQTTConnected) ? MQTTStatusConnected: MQTTStatusDisconnectPending;
+        }
```

Comment:
Reasoning I applied:
If the check failed then it would be good that we will exit early, but if it passed then the Mutex will be held for a longer period of time, unnecessarily.

But yes one thing can be done:
Take the mutex, do the check in the beginning, if fails, change status and bubble it up, else relinquish the mutex and take it back when needed again. Is this what we want to do?